### PR TITLE
Overhaul parsers/validators

### LINKED ISSUE #3: Revisit parsing/validation
#1 introduces parsers, and the main branch already has validators. 

I'm not currently sold on how these work - validators can only check the data, and "assert" that it's an expected type, and parsers can modify data but I'm a little iffy on the fact that they directly mutate the original request object.

It might be cleaner to require a "schema first" approach, perhaps even specifically designing around `zod` (or doing a `zod`-first approach like [zod-fetch](https://github.com/mattpocock/zod-fetch) (expect an object you can call `schema.parse(data)` with to get parsed data)

with #2 to consider, it may even be worth moving the original `req` and `res` objects to the handlerApi (and marking res as optional, since it won't exist for route handlers)

something like `handler: (parsedData: { body, query, cookies }, handlerApi: { failWithCode, succeedWithCode, req, res, ...etc }) => {` would allow further abstraction (and thus enabling supporting Route Handlers with the same or similar syntax)

in order to support setting an expected type for the response data while still inferring parsed data (without something awkward like `alwaysMatch`), it may be necessary to use a double call approach for typescript users, i.e. `method<ReturnType>()({ parsers, handler })`.

### PR DESCRIPTION
**Pull request by EskiMojo14**
Fixes #3

### CODE REVIEW

File: website/docs/concepts/method.mdx [hunk 1]
```diff
@@ -131,4 +131,107 @@ const endpoint = createEndpoint({
...
+
+:::caution
+When using Typescript, it's required to use the double call syntax `method<ReturnType>()(definition)` to use parsers.
+
+This allows the `ReturnType` type to be provided explicitly, while inferring the parsed types.
+
+If you're using Javascript, this is not required.
+:::
+
+<!-- TODO: work out why the below recipe results in body being typed as `any`
```
Comment (timdouglas) [inline website/docs/concepts/method.mdx:183]: worth creating an issue for this if you haven't already?

**Review by timdouglas — COMMENTED** [review]
lgtm, just a couple of queries

File: src/lib/execute.ts [hunk 2]
```diff
@@ -81,12 +68,17 @@ export const executeDefinition = async <
...
+      ConditionalBool<DisableAuthentication, undefined, Authentication>,
+      ExtraApi
     > = {
-      succeedWithCode,
-      failWithCode,
+      req,
+      // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
+      body: (await parsers?.body?.(req.body, failWithCode, req)) ?? req.body,
+      query:
+        (await parsers?.query?.(req.query, failWithCode, req)) ?? req.query,
```
[the remark this answers is not in the document]
Comment (EskiMojo14) [reply src/lib/execute.ts:81]: unsafe assignment only complains about `any` - req.query is Record<string, string | string[]> so it's fine

File: website/docs/concepts/method.mdx (hunk 1 above)
Comment (EskiMojo14) [reply website/docs/concepts/method.mdx:183]: https://github.com/actual-experience/create-endpoint-factory/issues/6

**Pull request merged**